File: backend/matlab/probabilities.m
clear all; close all; clc

upper_limit = 1:60;
normal_probability = 50063860;
new_probability = [2, 107, 1081, 5281, 17562, 45210, 100838, 199081, 354803, 591173, 927581, 1384055, 1976930, 2696228, 3581509, 4651632, 5894065, 7249689, 8773751, 10469474, 12289539, 14244037, 16349622, 18542928, 20714369, 22889240, 25101500, 27213564, 29362334, 31401351, 33379306, 35274810, 37018843, 38748473, 40360034, 41744581, 43042564, 44193642, 45204175, 46028887, 46771570, 47428541, 47997055, 48474781, 48917107, 49245778, 49517556, 49751198, 49868329, 49966048, 50024717, 50044289, 50063860, 50063860, 50063860, 50063860, 50063860, 50063860, 50063860, 50063860];
new_probability_percentage = 100 * new_probability / normal_probability;
bar(upper_limit, new_probability)
figure
bar(upper_limit, new_probability_percentage)

File: backend/myproject/modules/probabilities.py
import math


def calculate_total_probability(first_y, second_y, third_y, fourth_y, fifth_y, sixth_y):
    # Loop for global probability in each upper_limit
    new_probability_array = []
    new_probability_percentage_array = []
    upper_limit_array = []
    probability_data = []
    for upper_limit in range(1, 61):
        # Calculate probabilities for each position
        position_weights = []
        for y in [first_y, second_y, third_y, fourth_y, fifth_y, sixth_y]:
            position_weights.append(calculate_probability(y, upper_limit))

        # Calculate total probability
        weighted_total_probability = 1
        for weight in position_weights:
            weighted_total_probability *= weight

        total_probability = weighted_total_probability

        normal_probability = round(math.comb(60, 6))
        new_probability = round(normal_probability *
                                weighted_total_probability)
        new_probability_array.append(new_probability)
        new_probability_percentage_array.append(
            100 * new_probability / normal_probability)
        upper_limit_array.append(upper_limit)

        # print_probabilities(*position_weights, weighted_total_probability,
        #                     total_probability, new_probability, normal_probability)
        probability_data = [
            upper_limit_array, new_probability_array, new_probability_percentage_array]
    return probability_data


def calculate_probability(array_y, upper_limit):
    frequency_sum = sum(array_y[:upper_limit])
    total_games = sum(array_y)
    probability = frequency_sum / total_games
    return probability


def print_probabilities(first_guess_probability, second_guess_probability, third_guess_probability,
                        fourth_guess_probability, fifth_guess_probability, sixth_guess_probability,
                        global_probability, total_probability, new_probability, normal_probability):
    # Print the probabilities
    print("Probability of guessing first number correctly: {:.6f}".format(
        first_guess_probability))
    print("Probability of guessing second number correctly: {:.6f}".format(
        second_guess_probability))
    print("Probability of guessing third number correctly: {:.6f}".format(
        third_guess_probability))
    print("Probability of guessing fourth number correctly: {:.6f}".format(
        fourth_guess_probability))
    print("Probability of guessing fifth number correctly: {:.6f}".format(
        fifth_guess_probability))
    print("Probability of guessing sixth number correctly: {:.6f}".format(
        sixth_guess_probability))
    print("Probability of guessing all numbers correctly: {:.6f}".format(
        global_probability))
    print("Total probability of guessing all numbers correctly: {:.6f}".format(
        total_probability))
    print("New probability of guessing all numbers correctly: {:d}".format(
        new_probability))
    print("Normal probability of guessing all numbers correctly: {:d}".format(
        normal_probability))
